keras/engine/training.py

```diff
@@ -1985,21 +1896,23 @@ def evaluate_generator(self, generator, steps,
         wait_time = 0.01
         all_outs = []
         batch_sizes = []
+        is_dataset = isinstance(generator, Dataset)
+        if not is_dataset and pickle_safe:
+            warnings.warn(
+                """Using a generator with `pickle_safe=True` may duplicate your data.
+                Please considers using the `keras.data.Dataset` object.""")
         enqueuer = None
 
         try:
-            enqueuer = GeneratorEnqueuer(generator, pickle_safe=pickle_safe)
+            if is_dataset:
+                enqueuer = OrderedEnqueuer(generator, pickle_safe=pickle_safe)
+            else:
+                enqueuer = GeneratorEnqueuer(generator, pickle_safe=pickle_safe, wait_time=wait_time)
             enqueuer.start(workers=workers, max_q_size=max_q_size)
+            output_generator = enqueuer.get()
 
             while steps_done < steps:
-                generator_output = None
-                while enqueuer.is_running():
-                    if not enqueuer.queue.empty():
-                        generator_output = enqueuer.queue.get()
-                        break
-                    else:
-                        time.sleep(wait_time)
-
+                generator_output = next(output_generator)
```

Comment:
What about failure cases? Example: #6928 it is possible only x, only y, neither x nor y, or a tuple of some other unexpected size gets returned. 

At a minimum, check the tuple size and throw an exception if it doesn't match expectations. There are probably other cases like this in this pull request, it might be worth double checking.